File: mavigobi/models_regression/4-Support_Vector_Regression_SVR_regression/support_vector_regression.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov 28 20:47:59 2023

@author: mavigobi
"""

# SUPPORT VECTOR REGRESSION (SVR)


# DATA PRE PROCESSING 

# How to import  libraries 
import numpy as np;# Mathematical tool
import matplotlib.pyplot as plt; # Graphics tool
import pandas as pd; # Read data

# Import dataset
dataset = pd.read_csv('Position_Salaries.csv');
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:,2].values

# IN THIS CASE, THE VARIABLES SHOULDN'T BE DIVIDED

# Overfitting problem / Split data
# Dataset: training set & test set
"""
from sklearn.model_selection import train_test_split # four variables
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0) # It is common to sacrifice 20% of the data for the test
"""

# Variable scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X = sc_X.fit_transform(X)
y = sc_y.fit_transform(y.reshape(-1,1)) 

# GENERAL REGRESSION MODEL
from sklearn.svm import SVR
regression = SVR(kernel = "rbf")
regression.fit(X,y)


# PREDICTION OF OWN REGRESSION MODEL
y_pred = sc_y.inverse_transform([regression.predict(sc_X.transform(np.array([[6.5]])))])
"""y_pred = regression.predict(sc_X.transform([[6.5]]))
y_pred = sc_y.inverse_transform(y_pred)"""

# VISUALITATION OF POLYNOMIAL REGRESSION MODEL
#X_grid = np.arange(min(X), max(X), 0.01, dtype= float) 
#X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = "blue")
plt.plot(X, regression.predict(X), color = "red")
plt.title("Modelo de Regresión")
plt.xlabel("Sueldo anual del empleado")
plt.ylabel("Nivel laboral del empleado")
plt.show()

File: mavigobi/models_regression/4-Support_Vector_Regression_SVR_regression/support_vector_regression.R
# SUPPORT VECTOR REGRESSION (SVR)
# R - mavigobi


# PRE PROCESSING DATA

# Import dataset
dataset = read.csv('Position_Salaries.csv')
# Not define  X and y variables
dataset = dataset[,2:3]

# Overfitting problem. 
# Import packages
# install.packages("caTools")
# library(caTools) # import library
# set.seed(123) # define seed
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)

# Variables scaling
# It's important remember that Dummy variables are not numbers because Dummy are factors.
# training_set[, 2:3] = scale(training_set[,2:3]) # not : because R assume all rows and columns
# testing_set[,2:3] = scale(testing_set[,2:3])


# SVR REGRESSION MODEL
# should create "regression" own regression mode
#install.packages("e1071")
library("e1071")
regression = svm(formula = Salary ~.,
                 data = dataset,
                 type = "eps-regression",
                 kernel = "radial")

  
# PREDICTION OF REGRESSION MODEL
data_predict = predict(regression, newdata = data.frame(Level = 6.5))


# DISPLAY OF REGRESSION MODEL
library(ggplot2)
# x_grid increase number of data
#x_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = "blue") +
  geom_line(aes(x = dataset$Level, y = predict(regression, newdata = data.frame(Level = dataset$Level))),
            color = "red")+
  ggtitle("Modelo de Regresión SVR")+
  xlab("Nivel del empleado") +
  ylab("Salario anual del empleado")